sockeye/training.py

```diff
@@ -228,40 +275,141 @@ def fit(self,
+                                            validation_iter)
                 break
 
             if self.config.max_samples is not None and self.state.samples >= self.config.max_samples:
                 logger.info("Maximum # of samples (%s) reached", self.config.max_samples)
                 if not checkpoint_up_to_date:
                     time_cost = time.time() - tic
-                    self._create_checkpoint(checkpoint_decoder, time_cost, train_iter, validation_iter)
+                    self._create_checkpoint(checkpoint_decoder, loss_prefixes, mt_steps,
+                                            source_lang, target_lang, time_cost, train_iter,
+                                            validation_iter)
                 break
 
-            did_grad_step = self._step(batch=train_iter.next())
-            checkpoint_up_to_date = checkpoint_up_to_date and not did_grad_step
+            step, step_variant = next(steps_iter)
+            if step == "MT":
+                loss_prefix = get_loss_prefix("MT", step_variant[0].name, step_variant[1].name)
+                if step_variant == (source_lang, target_lang):
+                    did_grad_step = self._mt_step(batch=train_iter.next(), loss_prefix=loss_prefix)
+                elif step_variant == (target_lang, source_lang):
+                    did_grad_step = self._mt_step(batch=train_iter.next().reverse(), loss_prefix=loss_prefix)
+                else:
+                    assert False
+
+                # TODO: how to define epochs with the presence of the mono iters? For now we'll just keep epochs = parllel data epochs...
+                if not train_iter.iter_next():
+                    self.state.epoch += 1
+                    train_iter.reset()
+            elif step == "MASS":
+                step_lang = step_variant
+                # TODO: support for monolingual data without a language!?
+                assert step_lang.name is not None
+                loss_prefix = f"MASS-{step_lang.name}"
+                mono_iter = mono_iters[step_lang]
+                did_grad_step = self._mass_step(batch=mono_iter.next(),
+                                                # TODO: remove the vocab (was just for print debugging...)
+                                                loss_prefix=loss_prefix)
+
+                if not mono_iter.iter_next():
+                    mono_iter.reset()
+            elif step == "BT":
+                # bubble up/expose parameters
+                sampled_bt = False
+                tagged_bt = False
+
+                step_lang, bt_src_lang, bt_trg_lang = step_variant
+                # TODO: support for monolingual data without a language!?
+                assert step_lang.name is not None
+                loss_prefix = f"BT-{step_lang.name}"
+                # TODO: how to get mono dat here!?
+                mono_iter = mono_iters[step_lang]._iter
```

Comment:
why is this line different than for MASS above? (376)

Reply:
It is because we do the masking inside the iterator and the `_iter` just gives the raw mono data without masking. In any case this is not nice and I'd like to refactor this. The reason I do the masking in the iterator is to be able to use threading there. I wanted to reevaluate whether this was actually necessary and if not pull out masking from data_io.